virtdbg/vmx/ept.c

```diff
@@ -51,30 +39,45 @@ err_t init_ept() {
 
 err_t ept_map(uintptr_t address) {
     err_t err = NO_ERROR;
-
     lock(&m_ept_lock);
 
-    ept_entry_t* curr = m_root_pa;
-    for (size_t i = EPT_LEVELS; i >= 0; i--) {
-        // get the current index
-        size_t index = ((address >> ((9 * (i - 1)) + 12)) & 0x1FF);
-
-        if (!curr[index].r) {
-            // the entry is not present, create it
-            void* new_frame = pallocz(4096);
-            CHECK_ERROR(new_frame != NULL, ERROR_OUT_OF_RESOURCES);
+	int pteIdx   = (((address) >> 12) & 0x1ff);
 
-            curr[index].frame = (uint64_t)new_frame >> 12;
-            curr[index].r = 1;
-            curr[index].w = 1;
-            curr[index].x = 1;
+    ept_entry_t* cur = (ept_entry_t*)m_root_pa;
+    for (int i = 4; i > 1; i--) {
+        size_t idx = (((address) >> (12 + 9 * (i - 1))) & 0x1ff);
+        if (!(cur[idx].r)) {
+            uintptr_t alloc = (uintptr_t)pallocz_aligned(0x1000, 0x1000);
+            CHECK_ERROR(alloc != 0, ERROR_OUT_OF_RESOURCES);
+            cur[idx].frame = (alloc >> 12);
+            cur[idx].r = 1;
+            cur[idx].w = 1;
+            cur[idx].x = 1;
+            cur = (ept_entry_t*)alloc;
+        } else {
+            cur = (ept_entry_t*)(((uintptr_t)cur[idx].frame) << 12);
         }
-
-        // get the next level
-        curr = (ept_entry_t*)((uintptr_t)curr[index].frame << 12);
     }
 
+    cur[pteIdx].frame = address >> 12;
+    cur[pteIdx].r = 1;
+    cur[pteIdx].w = 1;
+    cur[pteIdx].x = 1;
+    cur[pteIdx].mem_type = 6;
+
+    //invept the entire EPT, since doing so on a single address requires the VPCID which we don't need

```

Comment:
Maybe move this to a different function and prettify it. Also we probably want to optionally support vpcid if it's better for perf